### ISSUE TITLE
jfrog rt build-publish/collect-build-env  is not validating artifacts uploaded for the corresponding build

### ISSUE THREAD

**Issue by thaveedu** [jfrog/jfrog-cli#195]
A typical life cycle for a jfrog build consists of the below 3 commands in sequence ,

```
jfrog rt upload sample.deb "debian-release-local/pool/sample/"
jfrog rt build-collect-env sample_test 1234
jfrog rt build-publish sample_test 1234
```

- The first command uploads the artifact to the server and captures the data in a partial directory, Here there is a good possibility this step is repeated for various number of artifacts which is part of the same build.
- The second command collects  the build env and add it to the json payload.A sample payload is given below
- The third command uploads the collected build data json to the server .

The problem is the publish command is not validating the presense of artifact data before pushing it to the server . I am attaching the payloads of publish command with and without the artifact data , both of them is successful which i feel should not be the case .In ideal scenario the publish operation should validate the payload before it uploads that to the server, Or at-least we need an interface to know there is no artifact uploaded to the server for this build parameters . 

Payload with artifact data

```javascript
{
	"name": "sample_test",
	"number": "128",
	"agent": {
		"name": "jfrog-cli-go",
		"version": "1.16.2"
	},
	"buildAgent": {
		"name": "GENERIC",
		"version": "1.16.2"
	},
	"modules": [{
		"properties": {},
		"id": "sample_test",
		"artifacts": [{
			"name": "sample.deb",
			"sha1": "d3d5fb2333ef537412db9fe16e5631554d1882b9",
			"md5": "cd2ec03dd82cc44e8cd6f6af0dda7c92"
		}]
	}],
	"started": "2018-07-23T18:19:20.618+0530",
	"properties": {
		"buildInfo.env.Apple_PubSub_Socket_Render": "/private/tmp/com.apple.launchd.NRY1oX27X1/Render",
		"buildInfo.env.COLORFGBG": "7;0",
		"buildInfo.env.COLORTERM": "truecolor",
		"buildInfo.env.COMMAND_MODE": "unix2003",
		"buildInfo.env.HOME": "/Users/txxxx",
		"buildInfo.env.HTTPS_PROXY": "127.0.0.1:8080",
		"buildInfo.env.HTTP_PROXY": "127.0.0.1:8080",
		"buildInfo.env.ITERM_PROFILE": "Default",
		"buildInfo.env.ITERM_SESSION_ID": "w0t1p0:54D8E3C8-9A21-4FAB-9E40-13D204DD8FDC",
		"buildInfo.env.JAVA_HOME": "/Library/Java/JavaVirtualMachines/jdk1.8.0_171.jdk/Contents/Home/",
		"buildInfo.env.JFROG_CLI_OFFER_CONFIG": "0",
		"buildInfo.env.LC_CTYPE": "UTF-8",
		"buildInfo.env.LESS": "-R",
		"buildInfo.env.LOGNAME": "txxx",
		"buildInfo.env.LSCOLORS": "Gxfxcxdxbxegedabagacad",
		"buildInfo.env.PAGER": "less",
		"buildInfo.env.SHELL": "/bin/zsh",
		"buildInfo.env.SHLVL": "1",
		"buildInfo.env.SSH_AUTH_SOCK": "/private/tmp/com.apple.launchd.hgRajzlj0a/Listeners",
		"buildInfo.env.TERM": "xterm-256color",
		"buildInfo.env.TERM_PROGRAM": "iTerm.app",
		"buildInfo.env.TERM_PROGRAM_VERSION": "3.1.7",
		"buildInfo.env.TMPDIR": "/var/folders/_t/7lspnwjx0sxg7f731d6qrbv9b8yg6t/T/",
		"buildInfo.env.USER": "txxx",
		"buildInfo.env.XPC_FLAGS": "0x0",
		"buildInfo.env.XPC_SERVICE_NAME": "0",
		"buildInfo.env._": "/usr/local/bin/jfrog",
		"buildInfo.env.__CF_USER_TEXT_ENCODING": "0x568F3CDA:0x0:0x0",
		"buildInfo.env.no_proxy": "127.0.0.1, localhost"
	},
	"artifactoryPrincipal": "embedded-tools-rw"
}
```

Payload without artifact data 
```javascript

{
	"name": "sample_test",
	"number": "31037",
	"agent": {
		"name": "jfrog-cli-go",
		"version": "1.16.2"
	},
	"buildAgent": {
		"name": "GENERIC",
		"version": "1.16.2"
	},
	"started": "2018-07-19T13:11:49.477+0530",
	"properties": {
		"buildInfo.env.Apple_PubSub_Socket_Render": "/private/tmp/com.apple.launchd.NRY1oX27X1/Render",
		"buildInfo.env.COLORFGBG": "7;0",
		"buildInfo.env.COLORTERM": "truecolor",
		"buildInfo.env.COMMAND_MODE": "unix2003",
		"buildInfo.env.HOME": "/Users/txxx",
		"buildInfo.env.HTTPS_PROXY": "127.0.0.1:8080",
		"buildInfo.env.HTTP_PROXY": "127.0.0.1:8080",
		"buildInfo.env.ITERM_PROFILE": "Default",
		"buildInfo.env.JAVA_HOME": "/Library/Java/JavaVirtualMachines/jdk1.8.0_171.jdk/Contents/Home/",
		"buildInfo.env.JFROG_CLI_OFFER_CONFIG": "0",
		"buildInfo.env.LC_CTYPE": "UTF-8",
		"buildInfo.env.LESS": "-R",
		"buildInfo.env.LOGNAME": "txxxx",
		"buildInfo.env.LSCOLORS": "Gxfxcxdxbxegedabagacad",
		"buildInfo.env.PAGER": "less",
		"buildInfo.env.SHELL": "/bin/zsh",
		"buildInfo.env.SHLVL": "2",
		"buildInfo.env.SSH_AUTH_SOCK": "/private/tmp/com.apple.launchd.hgRajzlj0a/Listeners",
		"buildInfo.env.TERM": "xterm-256color",
		"buildInfo.env.TERM_PROGRAM": "iTerm.app",
		"buildInfo.env.TERM_PROGRAM_VERSION": "3.1.7",
		"buildInfo.env.TERM_SESSION_ID": "w0t1p0:54D8E3C8-9A21-4FAB-9E40-13D204DD8FDC",
		"buildInfo.env.TMPDIR": "/var/folders/_t/7lspnwjx0sxg7f731d6qrbv9b8yg6t/T/",
		"buildInfo.env.USER": "txxxx",
		"buildInfo.env.XPC_FLAGS": "0x0",
		"buildInfo.env.XPC_SERVICE_NAME": "0",
		"buildInfo.env._": "/usr/local/bin/jfrog",
		"buildInfo.env.__CF_USER_TEXT_ENCODING": "0x568F3CDA:0x0:0x0",
		"buildInfo.env.__PYVENV_LAUNCHER__": "/usr/local/bin/python3",
		"buildInfo.env.no_proxy": "127.0.0.1, localhost"
	},
	"artifactoryPrincipal": "embedded-tools-rw"
}
```

**Comment by eyalbe4** [comment]
@thaveedu,
I believe that the second payload does not include the artifacts data because the upload command hadn't been executed before the publish command. It is correct?
A scenario in which build info does not include artifacts can be valid. For example, a build can include only dependencies. Another scenario is for build to not publish or consume any artifacts.
I do not think we would like to enforce it.
Can you perhaps share more information about this scenario? Why is this validation required?

**Comment by thaveedu** [comment]
@eyalbe4  Yes you are right , there we no upload happened before the publish . Thanks for the clarification on the scenarios ,but what will be the use of an empty build data in artifactory? I am confused here.

We developed a wrapper CLI over the jfrog CLI to facilitate setting up default configurations specific for each teams in our org .We provide upload as a single command , but we composed the build-collect-env and build-publish as a single command since we felt both will be executed together always . When we test some scenarios we noted that it is possible to publish the build data without any artifact been uploaded. We also noted, if we run the jfrog rt build-publish **without collecting** the build env it is **failing** since there is no partials available for that build . I felt this should be the behaviour of the build-publish command even if we run the build-collect-env before to that.

To summarise i have given the below sequences one with the build-collect-env  and another without , the first one will always pass regardless of how many times you execute it, while the second one will fail without any upload prior to that.

Always pass sequence
```bash
jfrog rt build-collect-env sample_test 1234
jfrog rt build-publish sample_test 1234
```

While this fails without any upload prior to it.
```bash
jfrog rt build-publish sample_test 1234
```

If we have an interface in the jFrog CLI to check whether there is artifact data available for that build we can avoid these kind of void publish event , but i can't find anything in the documentation. Please shed some light if there is anyway to achieve this.

**Comment by eyalbe4** [comment]
Thanks for the clarification @thaveedu.
There's currently no command which provides information about a local (non published) build.
If there was such an API, how would you incorporate it in your flow?

**Comment by thaveedu** [comment]
@eyalbe4  Before running every publish command , I would like to check whether there is some artifact uploaded for that build .

**Comment by eyalbe4** [comment]
I accidentally came across this old conversation we had @thaveedu. I didn't mean to abandon it. Sorry :(
If you wish to see the content of an unpublished build-info, without publishing it, you can run the following command - 
```
jf rt bp [build-name] [build-number] --dry-run
```
